cub/device/dispatch/dispatch_merge_sort.cuh

```diff
@@ -0,0 +1,697 @@
+                                   ValueT>;
+
+    cudaError error = cudaSuccess;
+
+    if (num_items == 0)
+      return error;
+
+    do
+    {
+      // Get device ordinal
+      int device_ordinal = 0;
+      if (CubDebug(error = cudaGetDevice(&device_ordinal))) break;
+
+      // Get shared memory size
+      int max_shmem = 0;
+      if (CubDebug(error = cudaDeviceGetAttribute(&max_shmem, cudaDevAttrMaxSharedMemoryPerBlock, device_ordinal))) break;
+
+      int tile_size = MergePolicyT::ITEMS_PER_TILE;
+      OffsetT num_tiles = cub::DivideAndRoundUp(num_items, tile_size);
+
+      size_t block_sort_shmem_size = BlockSortAgentT::SHARED_MEMORY_SIZE;
+      size_t merge_shmem_size = MergeAgentT::SHARED_MEMORY_SIZE;
+
+      size_t temp_storage1 = (1 + num_tiles) * sizeof(OffsetT);
+      size_t temp_storage2 = num_items * sizeof(KeyT);
+      size_t temp_storage3 = num_items * sizeof(ValueT) * !KEYS_ONLY;
+      size_t temp_storage4 =
+        vshmem_size(max_shmem,
+                    cub::max<size_t>(block_sort_shmem_size, merge_shmem_size),
+                    num_tiles);
+
+      void*  allocations[4]      = {nullptr, nullptr, nullptr, nullptr};
+      size_t allocation_sizes[4] = {temp_storage1, temp_storage2, temp_storage3, temp_storage4};
+
+      if (CubDebug(error = AliasTemporaries(d_temp_storage,
+                                            temp_storage_bytes,
+                                            allocations,
+                                            allocation_sizes)))
+      {
+        break;
+      }
+
+      if (d_temp_storage == nullptr)
+      {
+        // Return if the caller is simply requesting the size of the storage allocation
+        break;
+      }
+
+      int num_passes = static_cast<int>(thrust::detail::log2_ri(num_tiles));
+      bool ping = !(1 & num_passes);
+
+      auto merge_partitions = (OffsetT *)allocations[0];
+      auto keys_buffer         = (KeyT *)allocations[1];
+      auto items_buffer      = (ValueT *)allocations[2];
+
+      char* vshmem_ptr = temp_storage4 > 0 ? (char*)allocations[3] : nullptr;
+
+      // Invoke DeviceReduceKernel
+      BlockSortLauncher<KeyIteratorT,
+                        ValueIteratorT,
+                        OffsetT,
+                        MergePolicyT,
+                        CompareOpT,
+                        KeyT,
+                        ValueT>
+        block_sort_launcher(num_tiles,
+                            vshmem_ptr ? 0 : block_sort_shmem_size,
+                            ping,
+                            d_keys,
+                            d_items,
+                            num_items,
+                            compare_op,
+                            stream,
+                            keys_buffer,
+                            items_buffer,
+                            vshmem_ptr);
+
+      block_sort_launcher.launch();
+
+      if (debug_synchronous && (CubDebug(error = SyncStream(stream)))) break;
```

Comment:
This should be cleaned up a bit to improve clarity for readability/maintainability -- always use braces and avoid one-line conditionals or using `&&` in a condition statement for flow control.

```
if (debug_synchronous)
{
  if (CubDebug(error = SyncStream(stream)))
  {
    break;
  }
}
```